File: TsetlinMachine.py
import numpy as np
import random

class Clause:
    def __init__(self, noOfFeatures, learningRate, maxState): #initialises a rule
        self.rand = random.random

        self.maxState = maxState #maxState must be even
        self.cutOff = maxState / 2

        self.noOfFeatures = noOfFeatures
        self.noOfLiterals = noOfFeatures * 2

        self.learningRate = learningRate
        self.forgettingRate = 1 - learningRate

        self.states = np.ones(self.noOfLiterals) * self.cutOff
        for i in range(self.noOfLiterals):
            if (self.rand() >= 0.5):
                self.states[i] += 1
            else:
                self.states[i] -= 1

    def getOutput(self, input): #gives clause output as 0 or 1 depending on input, a numpy array of booleanised value of features
        return np.logical_and.reduce(np.where(self.states > self.cutOff, input, np.ones(self.noOfLiterals)))

    def s1(self):
        return (self.rand() <= learningRate)
    
    def s2(self):
        return (self.rand() <= (1 - learningRate))

    def typeIFeedback(self, input):
        if (self.getOutput(input)): #if clause output is 1
            for i in range(self.noOfLiterals): #run through all the literals
                if(input[i]): #if i-th literal is 1
                    if(self.s2()): #if s2 is true
                        if(self.states[i] < maxState): #increment the i-th state if less than maxState
                            self.states[i] += 1
                else: #if i-th literal is 0
                    if(self.s1()): #if s1 is true
                        if(self.states[i] > 1): #decrement the i-th state if greater than 1
                            self.states[i] -= 1
        else: #if clause output is 0
            for i in range(self.noOfLiterals):
                if(self.s1()):
                    if(self.states[i] > 1): #decrement the i-th state if greater than 1
                        self.states[i] -= 1

    def typeIIFeedback(self, input):
        if (self.getOutput(input)):
            for i in range(self.noOfLiterals):
                if (input[i]):
                    pass
                else:
                    if (self.states[i] <= self.cutOff):
                        if (self.states[i] < maxState):
                            self.states[i] += 1
            
class TM:
    def __init__(self, noOfClauses, noOfFeatures, learningRate, maxState, T):
        self.rand = random.random
        self.noOfClauses = noOfClauses #should be even
        self.T = T

        self.clauses = []
        for i in range(noOfClauses):
            clause = Clause(noOfFeatures, learningRate, maxState)
            self.clauses.append(clause)

    def classSum(self, input):
        sum = 0

        for i in range(self.noOfClauses):
            output = self.clauses[i].getOutput(input)

            if (i % 2 == 0):
                sum += output
            else:
                sum -= output

        return sum
    
    def expectedClassTraining(self, input):
        class_sum = self.classSum(input)
        if (class_sum > self.T):
            class_sum = self.T
        elif (class_sum < -1 * self.T):
            class_sum = -1 * self.T
        
        comparator = (self.T - class_sum) / (2 * self.T)

        for i in range(self.noOfClauses):
            if (self.rand() <= comparator):
                if (i % 2 == 0): #positive polarity clause
                    self.clauses[i].typeIFeedback(input)
                else: #negative polarity clause
                    self.clauses[i].typeIIFeedback(input)

    def randomClassTraining(self, input):
        class_sum = self.classSum(input)
        if (class_sum > self.T):
            class_sum = self.T
        elif (class_sum < -1 * self.T):
            class_sum = -1 * self.T

        comparator = (self.T + class_sum) / (2 * self.T)
        
        for i in range(self.noOfClauses):
            if (self.rand() <= comparator):
                if (i % 2 == 0): #positive polarity clause
                    self.clauses[i].typeIIFeedback(input)
                else: #negative polarity clause
                    self.clauses[i].typeIFeedback(input) 

class Machine:
    def __init__(self, classes, noOfClauses, noOfFeatures, learningRate, maxState, T):
        self.rand = random.randrange
        self.classes = classes
        self.noOfTM = len(classes)

        self.TMs = []
        for i in range(self.noOfTM):
            tm = TM(noOfClauses, noOfFeatures, learningRate, maxState, T)
            self.TMs.append([tm, self.classes[i]])

    def train(self, input, class_):
        for i in range(self.noOfTM):
            if (self.TMs[i][1] == class_):
                expectedTM = self.TMs[i][0]

        randValue = self.rand(0, self.noOfTM)
        while (self.TMs[randValue][1] == class_):
            randValue = self.rand(0, self.noOfTM)
        
        randomTM = self.TMs[randValue][0]

        expectedTM.expectedClassTraining(input)
        randomTM.randomClassTraining(input)

    def classify(self, input):
        maxOutput = 0
        maxOutputIndex = 0

        for i in range(self.noOfTM):
            if (self.TMs[i][0].classSum(input) > maxOutput):
                maxOutput = self.TMs[i][0].classSum(input)
                maxOutputIndex = i
        
        return self.TMs[maxOutputIndex][1]

    def test(self, input, class_):
        if (self.classify(input) == class_):
            return True
        else:
            return False
    
    def testOnData(self, data): #data is an array of the form [[class1, [input1]], [class2, [input2]], ...]
        correct = 0
        incorrect = 0

        for i in data:
            if (self.test(i[1], i[0])):
                correct += 1
            else:
                incorrect += 1
        
        return (correct, incorrect)

    def trainOnData(self, data, epochs): #data is an array of the form [[class1, [input1]], [class2, [input2]], ...]
        for j in range(epochs):
            for i in data:
                self.train(i[1], i[0])
            testResults = self.testOnData(data)
            print(f"Completed epoch-{j + 1} with accuracy {testResults[0] * 100 / (testResults[0] + testResults[1])}")


class MNIST:
    def __init__(self, epochs, threshold, classes, noOfClauses, noOfFeatures, learningRate, maxState, T):
        self.machine = Machine(classes, noOfClauses, noOfFeatures, learningRate, maxState, T)

        self.epochs = epochs

        with open("sample_data/mnist_train_small.csv") as f:
            from csv import reader
            stream = reader(f)
            data = []
            for i in stream:
                input = np.array([(int(a) > threshold) for a in i[1:]])
                data.append([int(i[0]), (np.array([input,(input^True)]).T).reshape(noOfFeatures*2)])
        self.data = data

        with open("sample_data/mnist_train_small.csv") as f:
            from csv import reader
            stream = reader(f)
            dataTest = []
            for i in stream:
                input = np.array([(int(a) > threshold) for a in i[1:]])
                dataTest.append([int(i[0]), (np.array([input,(input^True)]).T).reshape(noOfFeatures*2)])
        self.testingData = dataTest
    
    def trainAndTest(self):
        self.machine.trainOnData(self.data, self.epochs)
        test = self.machine.testOnData(self.testingData)
        print(f"Success: {(test[0] * 100) / (test[0] + test[1])}, Failure: {(test[1] * 100) / (test[0] + test[1])}")


epochs = 10
threshold = 10
classes = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
noOfClauses = 200
noOfFeatures = 784
learningRate = 1/7.5
maxState = 400
T = 10

myMNIST = MNIST(epochs, threshold, classes, noOfClauses, noOfFeatures, learningRate, maxState, T)
myMNIST.trainAndTest()

File: TsetlinMachine.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <time.h>
#include <fstream>
#include <string>
#include <sstream>
using namespace std;

//global variables
int maxStates = 400;
int cutOff = maxStates / 2;
int s = 7.5;
int T = 10;
int threshold = 10;
int noOfClasses = 10;
int noOfEpochs = 10;
int clausesPerComponent = 200;
int noOfFeatures = 784;
int noOfLiterals = noOfFeatures * 2;

class Clause
{
    //pending
    //decide if incrementIndex, decrementIndex, s1, s2 should be inline

    //meanings
    //states is a vector of states of each literal (between 1 to maxStates)
    //object is a given input vector of type bool of literals for operations like calculating clause output and giving feedbacks

    private:
    vector<int> states;

    public:
    Clause()
    {
        //initialise states randomly as cutOff or cutOff + 1
        for (int i = 0; i < noOfLiterals; ++i)
        {
            states.push_back(cutOff + ((rand()) & 1));
        }
    }

    //increment and decrement functions
    void incrementIndex(int& i)
    {
        if (states[i] < maxStates)
        {
            ++states[i];
        }
    }

    void decrementIndex(int& i)
    {
        if (states[i] > 1)
        {
            --states[i];
        }
    }

    //clause output
    bool clauseOutput(vector<bool>& object)
    {
        for (int i = 0; i < noOfLiterals; ++i)
        {
            if (states[i] > cutOff)
            {
                if (object[i] == 0)
                {
                    return 0;
                }
            }
        }
        return 1;
    }

    //comparators
    bool s1()
    {
        float random = rand();
        random /= RAND_MAX;
        return (random <= (1.0 / s));
    }

    bool s2()
    {
        float random = rand();
        random /= RAND_MAX;
        return (random <= (1.0 - (1.0 / s)));
    }

    //feedbacks
    void typeIFeedback(vector<bool>& object)
    {
        if (clauseOutput(object)) //if clause output is 1
        {
            for (int i = 0; i < noOfLiterals; ++i)
            {
                if (object[i]) //if literal is 1
                {
                    if (s2()) //if s2() returns "yes"
                    {
                        incrementIndex(i);
                    }
                }
                else //if literal is 0
                {
                    if (s1()) //if s1() returns "yes"
                    {
                        decrementIndex(i);
                    }
                }
            }
        }
        else //if clause output is 0
        {
            for (int i = 0; i < noOfLiterals; ++i)
            {
                if (s1()) //if s1() returns "yes"
                {
                    decrementIndex(i);
                }
            }
        }
    }

    void typeIIFeedback(vector<bool>& object)
    {
        if (clauseOutput(object)) //if clause output is 1
        {
            for (int i = 0; i < noOfLiterals; ++i)
            {
                if (!(object[i])) //if literal is 0
                {
                    if (states[i] <= cutOff)
                    {
                        incrementIndex(i);
                    }
                }
            }
        }
    }
};

class TsetlinMachineComponent
{
    //pending
    //decide whether c1, c2 should be inline or not

    //meanings
    //clauses if a vector of type Claues, holding values of all clauses
    //object is a given input vector of type bool of literals for operations like calculating clause output and giving feedbacks
    
    private:
    vector<Clause> clauses;

    public:
    TsetlinMachineComponent() 
    {
        clauses.resize(clausesPerComponent);
    }

    //class sum
    int classSum(vector<bool>& object)
    {
        int class_sum = 0;
        for (int i = 0; i < clausesPerComponent; ++i)
        {
            if (i & 1) //i is odd, negative polarity clause
            {
                class_sum += (-1 * clauses[i].clauseOutput(object));
            }
            else //i is even, positive polarity clause
            {
                class_sum += clauses[i].clauseOutput(object);
            }
        }
        return class_sum;
    }

    //clip class sum
    float clippedClassSum(vector<bool>& object)
    {
        int class_sum = classSum(object);

        if (class_sum >= T)
        {
            return T;
        }
        else if (class_sum <= -1* T)
        {
            return -1 * T;
        }
        else
        {
            return class_sum;
        }
    }

    //comparators
    bool c1(float& clipped_class_sum)
    {
        float random = rand();
        random /= RAND_MAX;
        return (random <= ((T - clipped_class_sum))/(2 * T));
    }

    bool c2(float& clipped_class_sum)
    {
        float random = rand();
        random /= RAND_MAX;
        return (random <= ((T + clipped_class_sum))/(2 * T));
    }

    //clause trainings
    void expectedClassTraining(vector<bool>& object)
    {
        float clipped_class_sum = clippedClassSum(object);
        for (int i = 0; i < clausesPerComponent; ++i)
        {
            if (c1(clipped_class_sum)) //training to be given
            {
                if (i & 1) //i is odd, clause polarity is negative
                {  
                    clauses[i].typeIIFeedback(object);
                }
                else //i is even, clause polarity is positive
                {
                    clauses[i].typeIFeedback(object);
                }
            }
        }
    }

    void randomClassTraining(vector<bool>& object)
    {
        float clipped_class_sum = clippedClassSum(object);
        for (int i = 0; i < clausesPerComponent; ++i)
        {
            if (c2(clipped_class_sum)) //training to be given
            {
                if (i & 1) //i is odd, clause polarity is negative
                {  
                    clauses[i].typeIFeedback(object);
                }
                else //i is even, clause polarity is positive
                {
                    clauses[i].typeIIFeedback(object);
                }
            }
        }
    }
};

class TsetlinMachine
{
    //meanings
    //machine is a vector of TsetlinMachineComponents for each class, representing our Tsetlin Machine
    //object is a given input vector of type bool of literals for operations like calculating clause output and giving feedbacks
    //label is the known class of object, its value ranges from 0 to noOfClasses - 1
    //data is a vector of objects
    //labels is a vector of int containing label corresponding to each object

    private:
    vector<TsetlinMachineComponent> machine;

    public:
    TsetlinMachine()
    {
        machine.resize(noOfClasses);
    }

    //testing
    bool test(vector<bool>& object, int& label) //for single object, returns whether correctly classified or not
    {
        int maxSum = machine[0].classSum(object);
        int maxSumClass = 0;

        for (int i = 1; i < noOfClasses; ++i)
        {
            int class_sum = machine[i].classSum(object);
            if (class_sum > maxSum)
            {
                maxSum = class_sum;
                maxSumClass = i;
            }
        }

        return (maxSumClass == label);
    }

    float testMachine(vector<vector<bool>>& data, vector<int>& labels) //for all objects in data, returns accuracy of model
    {
        int data_size = data.size();

        long long int correct = 0;
        for (int i = 0; i < data.size(); ++i)
        {
            correct += test(data[i], labels[i]);
        }

        return ((correct * 100.0) / data_size);
    }

    //training
    void train(vector<bool>& object, int& label) //for single object
    {
        int random_class = rand() % noOfClasses;
        while (random_class == label)
        {
            random_class = rand() % noOfClasses;
        }

        machine[label].expectedClassTraining(object);
        machine[random_class].randomClassTraining(object);
    }

    void trainMachine(vector<vector<bool>>& data, vector<int>& labels) //for all objects in data train machine once 
    {
        int data_size  = data.size();

        for (int i = 0; i < data_size; ++i)
        {
            train(data[i], labels[i]);
        }

        cout << "Accuracy: " << testMachine(data, labels);
    }
};

void conversionToBinary(string& temp, vector<vector<bool>>& data, long long int& count, long long int& colCount)
{
    int value = stoi(temp);

    if (value > threshold)
    {
        data[count][colCount++] = 1;
        data[count][colCount++] = 0;
    }
    else
    {
        data[count][colCount++] = 0;
        data[count][colCount++] = 1;
    }
}

int main()
{
    clock_t start = clock(); //initialise the start time to measure program runtime

    srand(time(NULL)); //seed srand with time(NULL)

    //loads training data
    long long int noOfTrainingSamples = 20000;

    ifstream trainingData;
    trainingData.open("/media/virinder/New Volume/Nikhat/TsetlinMachine/sample_data/mnist_train_small.csv");

    vector<vector<bool>> training_data (noOfTrainingSamples, vector<bool>(noOfLiterals));
    vector<int> training_labels (noOfTrainingSamples);

    string temp;
    long long int count = 0;
    long long int colCount;
    while (getline(trainingData, temp, '\n'))
    {
        stringstream s(temp);

        //read first digit from a row and store in labels
        getline(s, temp, ',');
        training_labels[count] = stoi(temp);

        //read subsequent information and store in data as literals based on thresholding
        colCount = 0;
        while (getline(s, temp, ','))
        {
            conversionToBinary(temp, training_data, count, colCount);
        }

        ++count;
    }
    
    //loads testing data
    long long int noOfTestingSamples = 10000;
    ifstream testingData;
    testingData.open("/media/virinder/New Volume/Nikhat/TsetlinMachine/sample_data/mnist_test.csv");

    vector<vector<bool>> testing_data (noOfTestingSamples, vector<bool>(noOfLiterals));
    vector<int> testing_labels (noOfTestingSamples);

    count = 0;
    while (getline(testingData, temp, '\n'))
    {
        stringstream s(temp);

        //read first digit from a row and store in labels
        getline(s, temp, ',');
        testing_labels[count] = stoi(temp);

        //read subsequent information and store in data as literals based on thresholding
        colCount = 0;
        while (getline(s, temp, ','))
        {
            conversionToBinary(temp, testing_data, count, colCount);
        }

        ++count;
    }

    //creates a tsetlin machine to classify MNIST data set
    TsetlinMachine MNIST;

    //train the machine and print accuracy during each epoch
    for (int i = 1; i <= noOfEpochs; ++i)
    {
        cout << "Executing epoch number " << i << ": ";
        MNIST.trainMachine(training_data, training_labels);
        cout << endl;
    }

    //test machine and print accuracy
    cout << "Completed all epochs, now testing the machine on test data." << endl;
    cout << "Accuracy on test data is " << MNIST.testMachine(testing_data, testing_labels);
    cout << endl;

    clock_t end = clock(); //initialise the end time to meausure program runtime

    //calculate time taken to run the program and print to std::cout
    double timeTaken = double(end - start) / CLOCKS_PER_SEC;
    cout << "Time taken (in seconds) is " << timeTaken << endl; 

    //debugging
    //
    // cout << endl;
    // for (auto& i : training_labels)
    // {
    //     cout << i << " ";
    // }
    // return 0;
    //
    // cout << endl;
    // cout << testing_labels[9999];
    // for (int i = 0; i < noOfLiterals; i += 2)
    // {
    //     if (i % 56 == 0)
    //     {
    //         cout << endl;
    //     }
    //     cout << testing_data[9999][i];
    // }
    // cout << endl;
    // cout << training_labels[0];
    // for (int i = 0; i < noOfLiterals; i += 2)
    // {
    //     if (i % 56 == 0)
    //     {
    //         cout << endl;
    //     }
    //     cout << training_data[0][i];
    // }
}